### PR TITLE
LPv2 writer

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:

This PR introduces three significant things:
- lp_v2 writer (`pyomo.repn.pluvins.lp_writer`): complete overhaul of the LP writer that is 30-40% faster
  - the new writer *will* complain about unrecognized active components (no more silently ignoring Disjunctions!)
  - the new writer will only emit the `ONE_VAR_CONSTANT` when it is required and will never emit the `c_e_ONE_VAR_CONSTANT` constraint (this lead to a lot of LP file baseline updates)
- LinearRepn walker (`pyomo.repn.linear`): new expression compiler, roughly equivalent to `generate_standard_repn(compute_values=True, quadratic=False)`
- QuadraticRepn walker (`pyomo.repn.quadratic`): new expression compiler, roughly equivalent to `generate_standard_repn(compute_values=True, quadratic=True)`

It also adds a `lp_diff` library for more robust comparisons of LP files (similar in structure to the `nl_diff` library).  Both differs have been moved up to `pyomo.repn.tests`.

**This PR activated the new writer as the default (and just updates all the baselines) in order to get full test coverage.  If desired, we can go back and preserve multiple baselines to allow for a transition period.**

Additional testing for the walkers will be forthcoming, but I wanted to get the PR up so people could start looking at it.

Performance:

image

## Changes proposed in this PR:
- see above

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by michaelbynum — COMMENTED** [review]
Lots of great things in this PR! I really appreciate that you are checking for unrecognized active components. 

Any plans to include an option like `compute_values=False` from `generate_standard_repn` in `LinerRepnVisitor` and `QuadraticRepnVisitor`? It would be nice to be able to remove `generate_standard_repn` from the code base so we don't have two walkers that do (nearly) the same thing.

File: pyomo/repn/plugins/lp_writer.py
Comment (jsiirola) [reply pyomo/repn/plugins/lp_writer.py:95]: Alas, for compatibility with the v1 writer.  Noth nlv1 and lpv2 default that to False, so the v2 writers do, too.  I would love to change that default later, though

File: pyomo/repn/plugins/lp_writer.py
Comment (jsiirola) [reply pyomo/repn/plugins/lp_writer.py:165]: It would write out unused variables to the LP file (with their bounds).  Since the v2 writer never goes and collects variables, it can't (easily or reliably) support this "feature".

File: pyomo/repn/plugins/lp_writer.py
Comment (jsiirola) [reply pyomo/repn/plugins/lp_writer.py:296]: Maybe, but it it is way easier to always create the Var object and then only emit it to the LP file if it is needed.  Note that it is never attached to the model - this is just an internal placeholder used by the writer (and if emitted, held by the SymbolMap)

File: pyomo/repn/linear.py
Comment (jsiirola) [reply pyomo/repn/linear.py:447]: My thought is that instead of complicating these walkers with more options, we should derive new walkers that replace specific handlers

**Comment by jsiirola** [comment]
> Any plans to include an option like `compute_values=False` from `generate_standard_repn` in `LinerRepnVisitor` and `QuadraticRepnVisitor`? It would be nice to be able to remove `generate_standard_repn` from the code base so we don't have two walkers that do (nearly) the same thing.

My plan it to create new walkers to handle the `compute_variables=False` option using the same pattern that the quadratic walker uses: derive from the Linear walker and override specific handlers (deferred to later PRs ... this one was already big enough).

File: pyomo/repn/util.py
Comment (jsiirola) [reply pyomo/repn/util.py:154]: Done

File: pyomo/repn/plugins/lp_writer.py
Comment (michaelbynum) [reply pyomo/repn/plugins/lp_writer.py]: Ahh, interesting... Good points.

File: pyomo/repn/plugins/lp_writer.py
Comment (michaelbynum) [reply pyomo/repn/plugins/lp_writer.py:95]: Makes sense. What would it take to change that default? Pyomo 7? Just curious.

File: pyomo/repn/plugins/lp_writer.py
Comment (michaelbynum) [reply pyomo/repn/plugins/lp_writer.py:165]: Got it.

File: pyomo/repn/util.py
Comment (michaelbynum) [reply pyomo/repn/util.py:154]: Thanks!

Comment (michaelbynum) [reply pyomo/repn/util.py:154]: Why don't I see it? Am I crazy?

File: pyomo/repn/linear.py
Comment (michaelbynum) [reply pyomo/repn/linear.py:447]: Great point

**Comment by michaelbynum** [comment]
> > Any plans to include an option like `compute_values=False` from `generate_standard_repn` in `LinerRepnVisitor` and `QuadraticRepnVisitor`? It would be nice to be able to remove `generate_standard_repn` from the code base so we don't have two walkers that do (nearly) the same thing.
> 
> My plan is to create new walkers to handle the `compute_variables=False` option using the same pattern that the quadratic walker uses: derive from the Linear walker and override specific handlers (deferred to later PRs ... this one was already big enough).

Great plan!

**Review by michaelbynum — APPROVED** [review]

**Comment by michaelbynum** [comment]
approved pending passing tests. Great work, John!

File: pyomo/repn/plugins/lp_writer.py
Comment (jsiirola) [reply pyomo/repn/plugins/lp_writer.py:95]: I think we can just change the default by fiat.  I was trying to minimize changes here, though, so deferred it to a subsequent PR.

File: pyomo/repn/quadratic.py
Comment (emma58) [inline pyomo/repn/quadratic.py:340]: This is SOOOO pretty! :)

File: pyomo/repn/plugins/lp_writer.py
Comment (emma58) [inline pyomo/repn/plugins/lp_writer.py:54]: Just thinking out loud, but we really should fix this soon... We hack around it a lot

File: pyomo/repn/plugins/lp_writer.py
Comment (emma58) [inline pyomo/repn/plugins/lp_writer.py:300]: My favorite 10 lines of code in this whole PR! :D

File: pyomo/repn/plugins/lp_writer.py
Comment (emma58) [inline pyomo/repn/plugins/lp_writer.py:392]: Should this be ignored? Or maybe it would be polite to at least log something suggesting that there might be a modeling error? I'm just thinking the most common cause of this is a typo, right? Or a bug in a transformation or something.

**Review by emma58 — COMMENTED** [review]
This is very happy, John! :) A few questions in the comments...

File: examples/pyomobook/scripts-ch/sudoku/sudoku_run.txt
Comment (jsiirola) [reply examples/pyomobook/scripts-ch/sudoku/sudoku_run.txt:2]: No - that struck me as a silly warning.  No objective is the same as constant objective of 0 (or for that matter, any other constant value.

File: pyomo/repn/quadratic.py
Comment (mrmundt) [inline pyomo/repn/quadratic.py:1]: A lot of this file doesn't seem to be covered by tests. Don't let it hold up the PR, but it makes me worry.

File: pyomo/repn/util.py
Comment (mrmundt) [reply pyomo/repn/util.py:198]: Gasp. Shame on the linter.

File: pyomo/repn/linear.py
Comment (jsiirola) [reply pyomo/repn/linear.py]: Actually, at this point, multiplier *can* be 0.  And that's OK.

File: pyomo/core/base/label.py
Comment (jsiirola) [reply pyomo/core/base/label.py:165]: Invisible to users.  Most *developers* don't even know about labalers

File: pyomo/contrib/preprocessing/plugins/induced_linearity.py
Comment (jsiirola) [reply pyomo/contrib/preprocessing/plugins/induced_linearity.py:105]: I think so: we want to transform from the leaves up

File: pyomo/repn/tests/test_linear.py
Comment (jsiirola) [reply pyomo/repn/tests/test_linear.py]: We map `None` to `nan` so that we can continue to process the expression (incase the subexpression is eventually multiplied by `0`).

**Pull request merged**
